voting/v2/contracts/Voting.sol

```diff
@@ -0,0 +1,654 @@
+			uint256 lockedFNFTBalance = getCurrentLockedFNFT(owner, _lockFarm);
+			if (lockedFNFTBalance > 0) {
+				flag = true;
+				break;
+			}
+		}
+		return flag;
+	}
+
+	// Check Locked FNFT in voting
+	function checkVotedFNFT(address owner) internal view returns (bool _flag) {
+		bool flag = false;
+		for (uint256 i = 0; i < getFarmsLength(); i++) {
+			LockFarm _lockFarm = getFarmsByIndex(i);
+			FNFT fNFT = FNFT(fnft[_lockFarm]);
+			if (canVoteByFNFT[fNFT]) {
+				FNFTInfo[] memory fInfo = _lockFarm.getFnfts(owner);
+				for (uint256 j = 0; j < fInfo.length; j++) {
+					uint256 lastVoted = lastVotedByFNFT[fNFT][fInfo[j].id]; // time of the last voted
+					uint256 time = block.timestamp - lastVoted;
+					if (time > voteDelay && fInfo[j].amount > 0) {
+						flag = true;
+						break;
+					}
+				}
+			}
+		}
+		return flag;
+	}
+
+	// Check Locked FNFT in voting
+	function getCurrentLockedFNFT(address owner, LockFarm _lockFarm)
+		internal
+		view
+		returns (uint256 _lockedFNFBalance)
+	{
+		uint256 lockedFNFTBalance = 0;
+
+		// Check FNFT Balance
+		FNFT fNFT = fnft[_lockFarm];
+		if (canVoteByFNFT[fNFT]) {
+			FNFTInfo[] memory fInfo = _lockFarm.getFnfts(owner);
+			for (uint256 i = 0; i < fInfo.length; i++) {
+				lockedFNFTBalance += fInfo[i].amount;
+			}
+		}
+		return lockedFNFTBalance;
+	}
+
+	// Compare Text
+	function compareStringsbyBytes(string memory s1, string memory s2) internal pure returns (bool) {
+		return keccak256(abi.encodePacked(s1)) == keccak256(abi.encodePacked(s2));
+	}
+
+	// Get weight for voting
+	function getWeightByUser(FNFT _fnft, uint256[] memory _fnftIds) internal view returns (uint256) {
+		uint256 totalWeightByUser = 0;
+		uint256 weightByFNFT = 0;
+
+		// Calculate of FNFT weight if there is FNFT voted
+		if (
+			address(_fnft) != address(0) &&
+			compareStringsbyBytes(_fnft.symbol(), 'HFNFT') &&
+			_fnftIds.length > 0
+		) {
+			for (uint256 j = 0; j < _fnftIds.length; j++) {
+				uint256 _lockedAmount = tokenVault.getFNFT(_fnftIds[j]).depositAmount;
+				IERC20Upgradeable _erc20Token = IERC20Upgradeable(tokenVault.getFNFT(_fnftIds[j]).asset);
+				uint256 calcAmount = convertToHEC(address(_erc20Token), _lockedAmount);
+				weightByFNFT += calcAmount;
+			}
+		}
+		totalWeightByUser = weightByFNFT;
+
+		return totalWeightByUser;
+	}
+
+	// Return HEC amount of the ERC20 token converted
+	function convertToHEC(address _stakingToken, uint256 _amount) public view returns (uint256) {
+		uint256 hecWeight = 0;
+		// Can input token can be weights
+		ERC20Upgradeable sToken = ERC20Upgradeable(address(_stakingToken));
+		if (
+			_stakingToken != address(0) &&
+			compareStringsbyBytes(stakingTokenSymbol[address(_stakingToken)], sToken.symbol())
+		) {
+			IERC20Upgradeable _token = IERC20Upgradeable(_stakingToken);
+			if (
+				address(lockFarmByERC20[_token]) != address(0) &&
+				(address(_stakingToken) == address(HECUSDC) || address(_stakingToken) == address(HECTOR))
```

Comment:
@xiangdev1016 I am concerned about this line because it limits the voting contract for future growth. What if next month, we will add HEC-BUSD pair on BSC chain. I think you may need to generalize/factor the LP factory contract. In other words, while we may use Spooky factory contract, we may not use it on other chains. 
The idea is how to make the Voting contract future-proof for LP pair.

Reply:
The PriceOracleAggregator is also useful on the voting contract, @xiangdev1016 @michaelnguyen102 ?

Reply:
@bejaxer It may work with some adjustments. Currently the PriceOracleAggregator contract exposes the priceInUSD kinda method while @xiangdev1016's contract needs to get the amount of HEC from the underlying LP pair.